core/src/main/java/org/apache/iceberg/ReplaceManifests.java

```diff
@@ -0,0 +1,243 @@
+  private final List<ManifestFile> newManifests = Collections.synchronizedList(new ArrayList<>());
+  private final Set<ManifestFile> processedManifests = Collections.synchronizedSet(new HashSet<>());
+  private final Map<Object, WriterWrapper> writers = Collections.synchronizedMap(new HashMap<>());
+  private final AtomicInteger manifestCount = new AtomicInteger(0);
+
+  private Function<DataFile, Object> clusterByFunc;
+  private Predicate<ManifestFile> predicate;
+
+  ReplaceManifests(TableOperations ops) {
+    super(ops);
+    this.ops = ops;
+    this.spec = ops.current().spec();
+    this.manifestTargetSizeBytes =
+      ops.current().propertyAsLong(MANIFEST_TARGET_SIZE_BYTES, MANIFEST_TARGET_SIZE_BYTES_DEFAULT);
+  }
+
+  @Override
+  protected String operation() {
+    return DataOperations.REPLACE;
+  }
+
+  @Override
+  public RewriteManifests set(String property, String value) {
+    summaryBuilder.set(property, value);
+    return this;
+  }
+
+  @Override
+  protected Map<String, String> summary() {
+    return summaryBuilder.build();
+  }
+
+  @Override
+  public ReplaceManifests clusterBy(Function<DataFile, Object> func) {
+    this.clusterByFunc = func;
+    return this;
+  }
+
+  @Override
+  public ReplaceManifests rewriteIf(Predicate<ManifestFile> pred) {
+    this.predicate = pred;
+    return this;
+  }
+
+  @Override
+  public List<ManifestFile> apply(TableMetadata base) {
+    Preconditions.checkNotNull(clusterByFunc, "clusterBy function cannot be null");
+
+    List<ManifestFile> currentManifests = base.currentSnapshot().manifests();
+
+    if (requiresRewrite(currentManifests)) {
+      // run the rewrite process
+      performRewrite(currentManifests);
+    } else {
+      // just keep any new manifests that were added since the last apply(), don't rerun
+      addExistingFromNewCommit(currentManifests);
+    }
+
+    // put new manifests at the beginning
+    List<ManifestFile> apply = new ArrayList<>();
+    apply.addAll(newManifests);
+    apply.addAll(existingManifests);
+
+    return apply;
+  }
+
+  private boolean requiresRewrite(List<ManifestFile> currentManifests) {
+    if (processedManifests.size() == 0) {
+      // nothing yet processed so perform a full rewrite
+      return true;
+    }
+    // if any processed manifest is not in the current manifest list, perform a full rewrite
+    Set<ManifestFile> set = Sets.newHashSet(currentManifests);
+    return processedManifests.stream().anyMatch(manifest -> !set.contains(manifest));
+  }
+
+  private void addExistingFromNewCommit(List<ManifestFile> currentManifests) {
+    // keep any existing manifests as-is that were not processed
+    existingManifests.clear();
+    currentManifests.stream()
+      .filter(manifest -> !processedManifests.contains(manifest))
+      .forEach(manifest -> existingManifests.add(manifest));
+  }
+
+  private void reset() {
+    existingManifests.clear();
+    processedManifests.clear();
+    newManifests.clear();
```

Comment:
By clearing `newManifests`, this loses track of any any manifests that were written by a previous call to `performRewrite`. Those should be deleted instead. Could you run `cleanAll` at the start of this method?

Reply:
Yes, I added this.